app/EventHandlers/TrackBannerDonationRedirects.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\App\EventHandlers;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+class TrackBannerDonationRedirects implements EventSubscriberInterface {
+
+	public const PIWIK_CAMPAIGN = 'piwik_campaign';
+	public const PIWIK_KWD = 'piwik_kwd';
+
+	private SessionInterface $session;
+	private string $submissionRoute;
+	private string $confirmationRoute;
+	private string $bannerSubmissionUrlParameter;
+
+	public static function getSubscribedEvents() {
+		return [
+			KernelEvents::REQUEST => 'onKernelRequest',
+			KernelEvents::RESPONSE => 'onKernelResponse'
+		];
+	}
+
+	public function __construct( SessionInterface $session, string $submissionRoute, string $confirmationRoute, string $bannerSubmissionUrlParameter ) {
+		$this->session = $session;
+		$this->submissionRoute = $submissionRoute;
+		$this->confirmationRoute = $confirmationRoute;
+		$this->bannerSubmissionUrlParameter = $bannerSubmissionUrlParameter;
+	}
+
+	public function onKernelRequest( RequestEvent $event ): void {
+		$request = $event->getRequest();
+
+		switch ( $request->attributes->get( '_route' ) ) {
+			case $this->submissionRoute:
+				$this->storeCampaignParameters( $request );
+				break;
+			case $this->confirmationRoute:
+				$this->restoreCampaignParameters( $event, $request );
+				break;
+		}
+	}
+
+	public function onKernelResponse( ResponseEvent $event ): void {
+		$request = $event->getRequest();
+
+		if ( $request->attributes->get( '_route' ) !== $this->submissionRoute ) {
+			$this->session->remove( self::PIWIK_CAMPAIGN );
+			$this->session->remove( self::PIWIK_KWD );
+		}
+	}
+
+	private function storeCampaignParameters( Request $request ): void {
+		if ( !$request->get( $this->bannerSubmissionUrlParameter ) ) {
```

Comment:
The methods are all aptly named and make this handler very readable. Maybe we can turn it up a notch.

```suggestion
		if ( $this->requestWasSubmittedFromBanner($request) ) {
```